File: Data Structures & Algorithms/longest-increasing-path-in-matrix/submission-1.cpp
class Solution
{
public:
    vector<vector<int>> dp;
    int longestIncreasingPath(vector<vector<int>> &matrix)
    {
        int lowest = numeric_limits<int>::min();
        int rows = matrix.size();
        int cols = matrix[0].size();
        dp = vector<vector<int>>(rows, vector<int>(cols, -1));

        int ans = 0;
        for (int i = 0; i < rows; i++)
        {
            for (int j = 0; j < cols; j++)
            {
                ans = max(ans, dfs(lowest, i, j, matrix));
            }
        }

        return ans;
    }

private:
    int dfs(int prev, int row, int col, vector<vector<int>> matrix)
    {
        int rows = matrix.size();
        int cols = matrix[0].size();

        if (row == rows || row == -1 || col == cols || col == -1)
        {
            return 0;
        }

        int curr = matrix[row][col];

        if (prev < matrix[row][col])
        {
            if (dp[row][col] != -1)
            {
                return dp[row][col];
            }
            int ans = 1 + max(dfs(curr, row + 1, col, matrix), max(dfs(curr, row - 1, col, matrix), max(dfs(curr, row, col + 1, matrix), dfs(curr, row, col - 1, matrix))));
            dp[row][col] = ans;
            return ans;
        }
        return 0;
    }
};

File: Data Structures & Algorithms/trapping-rain-water/submission-1.py
class Solution:
    def trap(self, height: List[int]) -> int:
        left_max = [0] * len(height)
        right_max = [0] * len(height)

        for i in range(len(height)):
            if i == 0:
                continue
            left_max[i] = max(left_max[i - 1], height[i - 1])
        for i in range(len(height) - 1, -1, -1):
            if i == len(height) - 1:
                continue
            right_max[i] = max(right_max[i + 1], height[i + 1])
        ans = 0
        for i in range(len(height)):
            surr = min(left_max[i], right_max[i])
            ans += (surr - height[i] if (surr - height[i]) > 0 else 0)
        return ans